core/vm/logger.go

```diff
@@ -18,12 +18,157 @@ package vm
 
 import (
 	"fmt"
+	"math/big"
 	"os"
 	"unicode"
 
 	"github.com/ethereum/go-ethereum/common"
 )
 
+type Storage map[common.Hash]common.Hash
+
+func (self Storage) Copy() Storage {
+	cpy := make(Storage)
+	for key, value := range self {
+		cpy[key] = value
+	}
+
+	return cpy
+}
+
+// StructuredLogCollector is the basic interface to capture emited logs by the EVM logger.
+type StructuredLogCollector interface {
+	// Adds the structured log to the collector.
+	AddStructuredLog(StructLog)
+}
+
+// LogCfg are the configuration options for structured logger the EVM
+type LogCfg struct {
+	DisableMemory  bool                   // disable memory capture
+	DisableStack   bool                   // disable stack capture
+	DisableStorage bool                   // disable storage capture
+	FullStorage    bool                   // show full storage (slow)
+	Collector      StructuredLogCollector // the log collector
+}
+
+// VmLogger is a logger interface used by the EVM to emit structured logs.
+type VmLogger interface {
+	// CaptureState caputeres the current EVM state and logs it.
+	CaptureState(pc uint64, op OpCode, gas, cost *big.Int, memory *Memory, stack *stack, contract *Contract, err error)
+}
```

Comment:
This interface is pointless. Users cannot supply their own implementation and they probably wouldn't want to (they'd just be reimplementing Logger).

I suggest that you 

* remove `VmLogger`

* unexport `Logger`

* make type of the `EVM.logger` field `*logger` and leave it nil unless `Config.Debug` is true

* change `EVM.Run` so it does

  ```go
  if evm.cfg.Debug {
      evm.logger.CaptureState(...)
  }
  ```
  
It will be less code overall.